internal/compiler/emitter.go

```diff
@@ -462,3 +469,93 @@ func getDeclarationDiagnostics(host EmitHost, file *ast.SourceFile) []*ast.Diagn
 	transform.TransformSourceFile(file)
 	return transform.GetDiagnostics()
 }
+
+type AnyProgram interface {
+	Options() *core.CompilerOptions
+	GetSourceFiles() []*ast.SourceFile
+	GetConfigFileParsingDiagnostics() []*ast.Diagnostic
+	GetSyntacticDiagnostics(ctx context.Context, file *ast.SourceFile) []*ast.Diagnostic
+	GetBindDiagnostics(ctx context.Context, file *ast.SourceFile) []*ast.Diagnostic
+	GetOptionsDiagnostics(ctx context.Context) []*ast.Diagnostic
+	GetProgramDiagnostics() []*ast.Diagnostic
+	GetGlobalDiagnostics(ctx context.Context) []*ast.Diagnostic
+	GetSemanticDiagnostics(ctx context.Context, file *ast.SourceFile) []*ast.Diagnostic
+	GetDeclarationDiagnostics(ctx context.Context, file *ast.SourceFile) []*ast.Diagnostic
+	Emit(ctx context.Context, options EmitOptions) *EmitResult
+}
+
+func HandleNoEmitOnError(ctx context.Context, program AnyProgram, file *ast.SourceFile) *EmitResult {
+	if !program.Options().NoEmitOnError.IsTrue() {
+		return nil // No emit on error is not set, so we can proceed with emitting
+	}
+
+	diagnostics := GetDiagnosticsOfAnyProgram(ctx, program, file, true, func(name string, start bool, nameStart time.Time) time.Time { return time.Time{} })
+	if len(diagnostics) == 0 {
+		return nil // No diagnostics, so we can proceed with emitting
+	}
+	return &EmitResult{
+		Diagnostics: diagnostics,
+		EmitSkipped: true,
+	}
+}
+
+func GetDiagnosticsOfAnyProgram(
+	ctx context.Context,
+	program AnyProgram,
+	file *ast.SourceFile,
+	skipNoEmitCheckForDtsDiagnostics bool,
+	recordTime func(name string, start bool, nameStart time.Time) time.Time,
+) []*ast.Diagnostic {
+	allDiagnostics := slices.Clip(program.GetConfigFileParsingDiagnostics())
+	configFileParsingDiagnosticsLength := len(allDiagnostics)
+
+	allDiagnostics = append(allDiagnostics, program.GetSyntacticDiagnostics(ctx, file)...)
+	allDiagnostics = append(allDiagnostics, program.GetProgramDiagnostics()...)
+
+	if len(allDiagnostics) == configFileParsingDiagnosticsLength {
+		// Options diagnostics include global diagnostics (even though we collect them separately),
+		// and global diagnostics create checkers, which then bind all of the files. Do this binding
+		// early so we can track the time.
+		bindStart := recordTime("bind", true, time.Time{})
```

Comment:
I don't know how I feel about this pattern... Feels a bit weird for the emitter file to be the one with this time tracking code, and then it be indiredcted through callbacks and so on, when it was quite direct before. What was wrong with the old way? Can this just return a function closure that is used to start/stop the range?